internal/watcher/watcher_test.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	t.Run("Count files", func(t *testing.T) {
+		assert.Equal(t, expectedFilesCount, len(files))
+	})
+
+	t.Run("Count directories", func(t *testing.T) {
+		assert.Equal(t, expectedDirCount, len(directories))
+	})
+
+	t.Run("Ignored extensions", func(t *testing.T) {
+		for _, f := range files {
+			if _, ok := ignoredExtensionsLookup[filepath.Ext(f)]; ok {
+				assert.Fail(t, "Haven't excluded files with ignored extensions.")
+				return
+			}
+		}
+	})
+}
+
+type mockWatchFuncProvider struct {
+	called      bool
+	calledMutex sync.Mutex
+}
+
+func (m *mockWatchFuncProvider) WatchFn(path string) error {
+	m.calledMutex.Lock()
+	m.called = true
+	m.calledMutex.Unlock()
+
+	return nil
+}
+
+func TestWatchDirectories(t *testing.T) {
+	mock := &mockWatchFuncProvider{}
+
+	baseDir, err := ioutil.TempDir("", "watcher_test1")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir: %v", err)
+	}
+
+	defer os.RemoveAll(baseDir)
+
+	file := filepath.Join(baseDir, "test_file")
+	if err = ioutil.WriteFile(file, []byte(":octopus:"), 0755); err != nil {
+		t.Fatalf("ioutil.WriteFile: %v", err)
+	}
+
+	_, directories, err := GetAllInDirectory(baseDir, []string{".swp"})
+	if err != nil {
+		t.Fatalf("GetAllInDirectory: %v", err)
+	}
+
+	done := make(chan struct{})
+
+	go func() {
+		for {
+			mock.calledMutex.Lock()
+			if mock.called {
+				mock.calledMutex.Unlock()
+				break
+			}
+			mock.calledMutex.Unlock()
+
+			if err := ioutil.WriteFile(file, []byte(":octopus: - :octopus:"), 0755); err != nil {
+				t.Errorf("ioutil.WriteFile: %v", err)
+			}
+
+			<-time.After(100 * time.Millisecond)
+		}
+
+		close(done)
+	}()
+
+	if err := WatchDirectories(directories, mock, done); err != nil {
+		t.Fatalf("WatchDirectories: %v", err)
+	}
+
+	assert.Equal(t, true, mock.called)
+}
+
+func TestNewDifferential(t *testing.T) {
+	t.Run("Empty struct", func(t *testing.T) {
+		expected := Differential{
+			ignoredExtensions: getIgnoredExtensions(nil),
+			previousChecksum:  make(map[string][16]byte),
+			previousContents:  make(map[string][]byte),
+			client:            nil,
+			url:               "",
+			token:             "",
+		}
+
+		actual, err := NewDifferential(nil, nil, nil, "", "")
+		if err != nil {
+			t.Fatalf("NewDifferential: %v", err)
+		}
+
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("Checksum & contents", func(t *testing.T) {
+		previousChecksum := make(map[string][16]byte)
+		previousContents := make(map[string][]byte)
+
+		baseDir, err := ioutil.TempDir("", "watcher_test1")
+		if err != nil {
+			t.Fatalf("ioutil.TempDir: %v", err)
+		}
+
+		defer os.RemoveAll(baseDir)
+
+		for i := 0; i < 10; i++ {
+			file := filepath.Join(baseDir, fmt.Sprintf("test_file_%d", i))
+			if err = ioutil.WriteFile(file, []byte(fmt.Sprintf(":octopus:%d", i)), 0755); err != nil {
+				t.Fatalf("ioutil.WriteFile: %v", err)
+			}
+
+			var b []byte
+			b, err = ioutil.ReadFile(file)
+			if err != nil {
+				t.Fatalf("ioutil.ReadFile: %v", err)
+			}
```

Comment:
```suggestion
			b, err := ioutil.ReadFile(file)
			if err != nil {
				t.Fatalf("ioutil.ReadFile: %v", err)
			}
```

Reply:
(this is not really important, so it's approved and up to you to apply).

Reply:
(this is not really important, so it's approved and up to you to apply).

Reply:
When done this way, govet reports a shadow declaration of err.